module/zfs/zfs_vnops.c

```diff
@@ -4001,6 +4105,37 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 
 	ZFS_EXIT(zfsvfs);
 	return (error);
+
+	/*
+	 * Clean-up path for broken link state.
+	 *
+	 * At this point we are in a (very) bad state, so we need to do our
+	 * best to correct the state. In particular, all of the nlinks are
+	 * wrong because we were destroying and creating links with ZRENAMING.
+	 *
+	 * In some form, all of thee operations have to resolve the state:
+	 *
+	 *  * link_destroy() *must* succeed. Fortunately, this is very likely
+	 *    since we only just created it.
+	 *
+	 *  * link_create()s are allowed to fail (though they shouldn't because
+	 *    we only just unlinked them and are putting the entries back
+	 *    during clean-up). But if they fail, we can just forcefully drop
+	 *    the nlink value to (at the very least) avoid broken nlink values
+	 *    -- though in the case of non-empty directories we will have to
+	 *    panic (otherwise we'd have a leaked directory with a broken ..).
+	 */
+commit_unlink_td_szp:
+	VERIFY3U(zfs_link_destroy(tdl, szp, tx, ZRENAMING, NULL), ==, 0);
+commit_link_tzp:
+	if (tzp) {
+		if (zfs_link_create(tdl, tzp, tx, ZRENAMING))
+			VERIFY3U(zfs_drop_nlink(tzp, tx, NULL), ==, 0);
+	}
+commit_link_szp:
+	if (zfs_link_create(sdl, szp, tx, ZRENAMING))
+		VERIFY3U(zfs_drop_nlink(szp, tx, NULL), ==, 0);
+	goto commit;
```

Comment:
@behlendorf When you take a look at this next can you please let me know whether `zfs_drop_nlink()` is actually needed and/or safe? My understanding of the link-related errors is that you shouldn't get one from `zfs_link_create` given that *we just* removed the link in this function. I also don't like that this will basically `BUG_ON` if we happen to hit this error with a non-empty directory...

Reply:
You should be able to use `zfs_dirent_lock` to lock the specific names to protect them.  Then you wouldn't need to call `zfs_drop_nlink` here, since it shouldn't be possible for these operations to fail.